h/api/views.py

```diff
@@ -156,15 +156,35 @@ def create(request):
     """Create an annotation from the POST payload."""
     json_payload = _json_payload(request)
 
+    # Validate the annotation for, and create the annotation in, PostgreSQL.
+    if request.feature('postgres'):
+        schema = schemas.CreateAnnotationSchema(request)
+        appstruct = schema.validate(copy.deepcopy(json_payload))
+        annotation = storage.create_annotation(request, appstruct)
+
+    # Validate the annotation for, and create the annotation in, Elasticsearch.
     legacy_schema = schemas.LegacyCreateAnnotationSchema(request)
     legacy_appstruct = legacy_schema.validate(copy.deepcopy(json_payload))
 
-    annotation = storage.create_annotation(request, legacy_appstruct)
+    # When 'postgres' is on make sure that annotations in the legacy
+    # Elasticsearch database use the same IDs as the PostgreSQL ones.
+    if request.feature('postgres'):
+        assert annotation.id
+        legacy_appstruct['id'] = annotation.id
 
-    _publish_annotation_event(request, annotation, 'create')
+    legacy_annotation = storage.legacy_create_annotation(request,
+                                                         legacy_appstruct)
```

Comment:
Right. So I think it only crashes when creating replies with the postgres feature flag is on. There's a mismatch because `fetch_annotation()` always fetches from either postgres or elasticsearch depending on the feature flag. But when the feature flag is on, the view code actually creates annotations in _both_ elasticsearch and postgres. When `legacy_create_annotation()` calls `fetch_annotation()` we always want to fetch from elastic, and when `create_annotation()` calls `fetch_annotation()` we always want to fetch from postgres, but when something else (e.g. the `read()` view) calls `fetch_annotation()` then we want to do the current behaviour of fetching from either postgres or elastic depending on the feature flag.

I guess that the annotation _should_ always exist in both postgres and es when the postgres feature flag is on. So one way to fix this might be to change the legacy transform code so that it works when the fetched parent annotation is a postgres one. It could access properties on the annotation classes, instead of reading them as dicts. That might have been simpler, but it didn't come to me first and also I don't really like changing the legacy code, so I went for this instead <https://github.com/hypothesis/h/pull/3153/commits/1a4f37d95caf8eb0d2a8d4d5f8c2a22206e21b8e>. This introduces some undesirable logic complexity but at least it can be removed when we remove the legacy es db.

Deleting and updating annotations are gonna need to do the same thing as well.